File: LeetCode/LeetCode 337. 打家劫舍 III (dp).py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        def dfs(root):
            if root is None:
                return [0, 0]
            l = dfs(root.left)
            r = dfs(root.right)
            sel = root.val + l[1] + r[1]
            not_sel = max(l[0], l[1]) + max(r[0], r[1])
            return [sel, not_sel]
    
        ans = dfs(root)
        return max(ans[0], ans[1])

File: LeetCode/LeetCode 337. 打家劫舍 III (dp).cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

struct Node {
    int sel, not_sel;
};

class Solution {
public:
    Node dfs(TreeNode* root) {
        if(!root) return {0, 0};
        auto l = dfs(root->left);
        auto r = dfs(root->right);
        int sel = root->val + l.not_sel + r.not_sel;
        int not_sel = max(l.sel, l.not_sel) + max(r.sel, r.not_sel);
        return {sel, not_sel};
    }

    int rob(TreeNode* root) {
        auto ans = dfs(root);
        return max(ans.sel, ans.not_sel);
    }
};